Pal 863 data service audit message (#238)
gchq/Palisade-services: data-service/README.md

<!---
Copyright 2018-2021 Crown Copyright

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
--->
# Data Service

<span style="color:red">**Note:** As noted in the [documentation root](../README.md) Palisade is in an early stage of development, therefore the precise APIs contained in these service modules are likely to adapt as the project matures.</span>

## Overview

The core API for the data service.

The responsibilities of the data service is to provide the client with access to the resource that was requested.  This data will have been filtered based the user and context.

**Note:** A resource could be a file, stream, directory or even the system

> resource (policies added to the system resource would be applied globally).

Reviewer: Do we not need some more information in this ReadMe? Like the request object as JSON that can be used for testing and what the response back would look like?
We should probably include some details about how it goes to Redis to get the authorised request.

Reviewer 2: I agree with @pd104923 that external APIs should be documented. These should include REST and Kafka based inputs/outputs (they should be the same). The documented JSON for these APIs should be used "as-is" in the unit tests.

Author: Have updated the README with the message type for the request, response and the audit message types.

<!---
Copyright 2018-2021 Crown Copyright

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
--->
# Data Service

<span style="color:red">**Note:** As noted in the [documentation root](../README.md) Palisade is in an early stage of development, therefore the precise APIs contained in these service modules are likely to adapt as the project matures.</span>

## Overview

> The core API for the data service accepts a request from the client and returns the resources that have been processed from the initial request for resources.  These resources will be filtered and redacted based on the processing of the request in the inital resource request.

Reviewer 3: ```suggestion
The core API for the Data Service accepts a request from the client and returns the resources that have been processed from the initial request for resources.  These resources will be filtered and redacted based on the processing of the request in the inital resource request.
```

The client request will contain the token and resource id that is used to uniquely identify this resource request.

## Message Model and Database Domain

| DataRequest     | (response to client)   | AuditSuccessMessage  | AuditErrorMessage |
|:----------------|:-----------------------|:---------------------|:------------------|

> | token           | OutputStream           | *token               | *token            | 

Reviewer 4: ```suggestion
| *token          | OutputStream           | *token               | *token            | 
```

| leafResourceId  |                        | leafResourceId       | leafResourceId    |  
|                 |                        | userId               | ***userId         |
|                 |                        | resourceId           | ***resourceId     |
|                 |                        | context              | ***context        | 
|                 |                        | **attributes         | ***attributes     |
|                 |                        | serverMetadata       | error             |
|                 |                        |                      | serverMetadata    |
  
*token is stored in the header metadata  

**attributes will include the numbers for records processed and records returned

***data that may not be available depending on when the error occurred

## REST Interface

The application exposes one endpoint to the client for retrieving the resources. This will be the data that has previously 
been requested and prepared in the initial request to Palisade services. 
* `POST api/read/chunked`

>     - returns a `202 ACCEPTED` and an OutputStream which will provides the resource.

Reviewer 4: ```suggestion
    - returns a `202 ACCEPTED` and a streamed HTTP response body which will provides the resource.
```

## Example JSON Request
```
curl -X POST data-service/read/chunked  -H "content-type: application/json" --data \
'{
   "token": "test-token",
   "leafResourceId": "test-leafResourceId"
 }'
```

> 

Reviewer 4: I would say "add an example response", but that is literally just the data, so...

Reviewer 3: added

Reviewer 3: added
